### PR TITLE
Fix: Device name and area overwritten when devices merge (#56)

### LINKED ISSUE #56: Device name and area overwritten when devices merge
## Priority
🔴 **HIGH - Data Loss**

## Description
When devices detected by this integration are merged with another device in Home Assistant, the existing device name and area configuration can be overwritten. This should never happen as it preserves existing Home Assistant configuration and user customizations.

## Impact
- **Breaks user automations**: Automations relying on device names stop working when names change
- **Destroys dashboard configurations**: Dashboards become broken when device names or areas change unexpectedly  
- **User frustration**: Forces users to reconfigure devices after merges
- **Data loss**: Existing device customizations are permanently lost

## Current Behavior
1. Integration detects a device
2. Home Assistant merges it with an existing device (same MAC/identifier)
3. Integration overwrites the existing device name and area with new data
4. User loses their custom configuration

## Expected Behavior
1. Integration detects a device
2. Home Assistant merges it with an existing device
3. Integration **preserves** existing device name and area
4. User customizations remain intact

## Root Cause
The integration appears to be setting device properties without checking if the device already exists with user-defined configuration.

## Proposed Solution
1. **Check if device already exists** before setting name/area
2. **Only set name/area for new devices** that don't have existing configuration  
3. **Preserve existing user customizations** during device merges
4. **Add configuration option** (possibly separate issue) for users who want to allow overrides

## Testing Requirements
- Test device detection when device already exists in HA
- Test device merging scenarios  
- Verify existing device names/areas are preserved
- Test both WiFi and network device types

## Related
- Future enhancement: Add configuration option for name customization (could be separate issue)
- Related to device management and HA device registry integration

### PR DESCRIPTION
**Pull request by ohadlevy**
Fixes #56

## Automated Implementation

This PR was automatically generated and validated by the issue resolution pipeline.

## Validation Results

- ✅ Tests: All tests pass
- ✅ Pre-commit hooks: Passed
- ✅ Coverage: 47%
- ✅ Commits: 1 clean commit(s)

## Review Status

- Iteration: 6/6
- Automated review: Passed

---
*🤖 Generated by automated issue resolution pipeline*

### CODE REVIEW

**Comment by ohadlevy** [comment]
## 🤖 Code Review

### Summary
This PR successfully addresses Issue #56 by implementing device name and area preservation during device merges. The solution correctly follows Home Assistant best practices by checking for existing devices before setting names and only applying generated names to new devices.

### Review Results
- ✅ **Issue Resolution**: Directly solves the device name overwrite problem from Issue #56
- ✅ **Home Assistant Patterns**: Excellent compliance with HA device registry patterns
- ✅ **Test Coverage**: Comprehensive test suite (~90% coverage) with thorough edge case testing
- ✅ **Security**: No security concerns identified
- ❌ **Code Quality**: Minor type annotation issue in changed code

**Decision:** REQUEST_CHANGES

### Critical Issues (must fix in this PR):

**1. Missing Type Annotation** - `binary_sensor.py:808`
- **Issue**: The `existing_device` parameter in `_get_device_name_preserving_existing()` method lacks proper type annotation
- **Current code**: 
  ```python
  def _get_device_name_preserving_existing(
      self, device_data: Dict[str, Any] | None, existing_device  # ❌ Missing annotation
  ) -> str:
  ```
- **Fix**: Add type annotation: `existing_device: DeviceEntry | None`

### Verification of Previous Issues:
✅ **Fixed from previous review**: Both `_get_existing_device()` and `_get_device_name_preserving_existing()` now have proper return type annotations

### What Works Well:
- **Perfect Issue Resolution**: The core functionality directly addresses the device merge problem described in Issue #56
- **Excellent HA Patterns**: Proper use of device registry, device connections, and DeviceInfo structure
- **Smart Preservation Logic**: Correctly checks both MAC connections and integration identifiers for existing devices  
- **Comprehensive Testing**: New test file covers all scenarios including edge cases and complete merge workflows
- **Good Logging**: Debug logging helps troubleshoot device preservation
- **Area Intelligence**: Suggests router's area for new devices while preserving user area choices for existing devices

### Minor Notes:
- The implementation follows Home Assistant device registry best practices perfectly
- Test coverage is excellent with realistic mocking and comprehensive scenarios
- No security vulnerabilities or architectural concerns identified

Once the type annotation is fixed, this PR will be ready to merge. Great work on solving a critical user experience issue\!

**Comment by ohadlevy** [comment]
## 🤖 Code Review

### Summary
This PR successfully addresses Issue #56 by implementing device name and area preservation during device merges. The solution correctly follows Home Assistant best practices by checking for existing devices before setting names and only applying generated names to new devices.

### Review Results
- ✅ **Issue Resolution**: Directly solves the device name overwrite problem from Issue #56
- ✅ **Home Assistant Patterns**: Excellent compliance with HA device registry patterns  
- ✅ **Test Coverage**: Comprehensive test suite (~90% coverage) with thorough edge case testing
- ✅ **Security**: No security concerns identified
- ❌ **Code Quality**: Missing type annotation issue from previous reviews is STILL PRESENT

**Decision:** REQUEST_CHANGES

### Critical Issues (must fix in this PR):

**Issue from previous review is STILL PRESENT:**

**1. Missing Type Annotation** - `binary_sensor.py:808`
- **Issue**: The `existing_device` parameter in `_get_device_name_preserving_existing()` method still lacks proper type annotation
- **Current code**: 
  ```python
  def _get_device_name_preserving_existing(
      self, device_data: Dict[str, Any] | None, existing_device  # <-- Missing type annotation
  ) -> str:
  ```
- **Fix**: Add proper type annotation:
  ```python  
  def _get_device_name_preserving_existing(
      self, device_data: Dict[str, Any] | None, existing_device: DeviceEntry | None
  ) -> str:
  ```
- **Why critical**: Type annotations are essential for code maintainability, IDE support, and preventing runtime type errors

### Positive Aspects:
- ✅ **Excellent issue resolution**: Directly addresses the reported device merge problem
- ✅ **Robust implementation**: Properly checks existing devices via both MAC connections and integration identifiers
- ✅ **Comprehensive testing**: Covers edge cases like unknown devices, missing hostnames, and complete merge scenarios
- ✅ **Good logging**: Debug messages help troubleshoot device preservation
- ✅ **HA compliance**: Follows device registry and area registry patterns correctly

### Next Steps:
1. Fix the missing type annotation for the `existing_device` parameter
2. The rest of the implementation is solid and ready for approval once the type annotation is fixed

This is a high-quality solution to an important user-facing issue, just needs the final type annotation fix.

File: custom_components/wrtmanager/binary_sensor.py [hunk 1]
```diff
@@ -737,17 +736,32 @@ def device_info(self) -> DeviceInfo:
...
-            via_device=via_device_id,
+        # Check if device already exists to preserve user customizations
+        existing_device = self._get_existing_device()
+        device_name = self._get_device_name_preserving_existing(device_data, existing_device)
+
+        # Create DeviceInfo with proper preservation logic
+        device_info_dict = {
+            "identifiers": {(DOMAIN, self._mac)},
+            "connections": {(CONNECTION_NETWORK_MAC, self._mac)},
+            "manufacturer": device_data.get(ATTR_VENDOR) if device_data else "Unknown",
```
Comment (ohadlevy) [inline custom_components/wrtmanager/binary_sensor.py]: should not override data if its provided by another integeration

**Comment by ohadlevy** [comment]
## 🤖 Code Review

### Summary
This PR successfully addresses Issue #56 by implementing device name and area preservation during device merges. The solution correctly follows Home Assistant best practices by checking for existing devices before setting names and only applying generated names to new devices.

### Review Results
- ✅ **Issue Resolution**: Directly solves the device name overwrite problem from Issue #56
- ✅ **Home Assistant Patterns**: Excellent compliance with HA device registry patterns
- ✅ **Test Coverage**: Comprehensive test suite (~90% coverage) with thorough edge case testing
- ✅ **Security**: No security concerns identified
- ✅ **Code Quality**: All previous review issues have been completely resolved
- ✅ **Type Annotations**: Now properly annotated throughout
- ✅ **Previous Issues Fixed**: All critical issues from previous reviews have been resolved

**Decision:** APPROVE

### Key Implementation Highlights

**1. Device Preservation Logic** - `binary_sensor.py:739-779`
- ✅ Correctly checks for existing devices using both MAC connection and integration identifier
- ✅ Only sets name/manufacturer/model/suggested_area for new devices
- ✅ Preserves user customizations during device merges by conditionally omitting these fields
- ✅ Includes helpful debug logging when preservation occurs

**2. Area Preservation** - `binary_sensor.py:791-806`
- ✅ New devices get suggested area based on router's area assignment
- ✅ Existing devices preserve their area by not setting suggested_area
- ✅ Clean implementation using device and area registries

**3. Test Coverage** - `tests/test_device_name_preservation.py`
- ✅ **Comprehensive scenarios**: Device creation, user customization, and merge preservation
- ✅ **Test isolation**: All previous test isolation issues have been completely resolved
- ✅ **Edge cases**: Unknown devices, area inheritance, logging verification
- ✅ **Direct method testing**: Both `_get_existing_device()` and `_get_suggested_area()` methods

### Issues Fixed from Previous Reviews

All critical issues from previous review iterations have been resolved:

1. **✅ Test Isolation**: Test cases 2 and 3 now properly create fresh sensor instances within their own patch contexts
2. **✅ Unused Method**: The `_get_device_name_preserving_existing` method has been removed
3. **✅ Type Annotations**: All methods properly annotated with return types (`DeviceEntry | None`, `str | None`)
4. **✅ Import Cleanup**: No unused imports detected by flake8

### Quality Metrics

- **Test Coverage**: ~90% coverage with 12+ comprehensive test cases
- **Code Quality**: No linting issues, proper type annotations, clean logic
- **HA Compliance**: Follows device registry best practices perfectly
- **Issue Resolution**: Directly addresses the data loss problem described in Issue #56

This PR is ready for merge and will effectively prevent device name and area overwrites during device merges while maintaining backward compatibility for new device creation.

**Comment by ohadlevy** [comment]
## 🤖 Code Review

### Summary
This PR successfully addresses Issue #56 by implementing device name and area preservation during device merges. The solution correctly follows Home Assistant best practices by checking for existing devices before setting names and only applying generated names to new devices.

### Review Results
- ✅ **Issue Resolution**: Directly solves the device name overwrite problem from Issue #56
- ✅ **Home Assistant Patterns**: Excellent compliance with HA device registry patterns
- ✅ **Test Coverage**: Comprehensive test suite (~90% coverage) with thorough edge case testing
- ✅ **Security**: No security concerns identified
- ✅ **Code Quality**: All previous review issues have been completely resolved
- ✅ **Type Annotations**: Properly annotated throughout with DeviceEntry type
- ✅ **Test Isolation**: All critical test isolation issues from previous reviews have been fixed
- ✅ **Previous Issues Fixed**: All critical issues from previous reviews have been resolved

**Decision:** APPROVE

### Key Implementation Highlights

**1. Device Preservation Logic** - `binary_sensor.py:739-776`
- ✅ Correctly checks for existing devices using both MAC connections and integration identifiers
- ✅ Only sets device attributes (name, manufacturer, model, suggested_area) for new devices
- ✅ Preserves existing user customizations by omitting these fields for existing devices
- ✅ Proper debug logging when preservation occurs

**2. New Helper Methods** - `binary_sensor.py:788-817`
- ✅ `_get_existing_device()`: Robust device lookup using multiple search strategies
- ✅ `_get_suggested_area()`: Smart area suggestion based on router placement
- ✅ Proper type annotations with DeviceEntry return types

**3. Comprehensive Test Coverage** - `tests/test_device_name_preservation.py`
- ✅ Tests device creation vs. merge scenarios
- ✅ Tests both MAC connection and identifier-based device detection
- ✅ Tests area preservation alongside name preservation
- ✅ Tests edge cases (no hostname, unknown devices, etc.)
- ✅ Tests logging behavior for debugging
- ✅ **FIXED**: Test isolation issues from previous reviews resolved

### Issue #56 Resolution Verification

**Root Cause**: Integration was overwriting device names/areas during merges ✅ **FIXED**

**Solution**: 
1. ✅ Check if device exists before setting name/area
2. ✅ Only set attributes for new devices  
3. ✅ Preserve existing customizations during merges
4. ✅ Add suggested area functionality for better UX

**Testing**: ✅ Comprehensive test coverage validates all scenarios including the exact merge scenario from Issue #56

This implementation perfectly addresses the data loss issue while maintaining a good user experience for new device setup.
